Source/Dafny/Translator.cs

```diff
@@ -8206,17 +8294,62 @@ private class WFOptions
+            PutSourceIntoLocal();
+            var toBound = Basetypes.BigNum.FromBigInt(BigInteger.One << toWidth); // 1 << toWidth
+            var bound = BplBvLiteralExpr(tok, toBound, expr.Type.AsBitVectorType);
+            var boundsCheck = FunctionCall(expr.tok, "lt_bv" + fromWidth, Bpl.Type.Bool, o, bound);
+            builder.Add(Assert(tok, boundsCheck,
+              string.Format("{0}bit-vector value to be converted might not fit in {1}", errorMsgPrefix, toType)));
+          }
+        } else if (expr.Type.IsBigOrdinalType) {
+          PutSourceIntoLocal();
+          var oi = FunctionCall(tok, "ORD#Offset", Bpl.Type.Int, o);
+          int toWidth = 16;
+          var toBound = Basetypes.BigNum.FromBigInt(BigInteger.One << toWidth); // 1 << toWidth
+          var bound = Bpl.Expr.Literal(toBound);
+          var boundsCheck = Bpl.Expr.Lt(oi, bound);
+          builder.Add(Assert(tok, boundsCheck,
+            string.Format("{0}ordinal value to be converted might not fit in {1}", errorMsgPrefix, toType)));
         }
-      } else if (toType.IsBigOrdinalType && expr.Type.IsNumericBased(Type.NumericPersuation.Int)) {
-        PutSourceIntoLocal();
-        Bpl.Expr boundsCheck = Bpl.Expr.Le(Bpl.Expr.Literal(0), o);
-        builder.Add(Assert(tok, boundsCheck, string.Format("{0}a negative integer cannot be converted to an {1}", errorMsgPrefix, toType)));
-      } else if (expr.Type.IsBigOrdinalType && toType.IsNumericBased(Type.NumericPersuation.Int)) {
-        PutSourceIntoLocal();
-        Bpl.Expr boundsCheck = FunctionCall(tok, "ORD#IsNat", Bpl.Type.Bool, o);
-        builder.Add(Assert(tok, boundsCheck, string.Format("{0}value to be converted might be bigger than every natural number", errorMsgPrefix)));
+      } else if (toType.IsBigOrdinalType) {
+          if (expr.Type.IsNumericBased(Type.NumericPersuation.Int)) {
+            PutSourceIntoLocal();
+            Bpl.Expr boundsCheck = Bpl.Expr.Le(Bpl.Expr.Literal(0), o);
+            builder.Add(Assert(tok, boundsCheck,
+              string.Format("{0}a negative integer cannot be converted to an {1}", errorMsgPrefix, toType)));
+          }
+
+          if (expr.Type.IsNumericBased(Type.NumericPersuation.Real)) {
+            PutSourceIntoLocal();
+            var oi = FunctionCall(tok, BuiltinFunction.RealToInt, null, o);
+            Bpl.Expr boundsCheck = Bpl.Expr.Le(Bpl.Expr.Literal(0), oi);
+            builder.Add(Assert(tok, boundsCheck,
+              string.Format("{0}a negative real cannot be converted to an {1}", errorMsgPrefix, toType)));
+          }
+          
+      } else if (toType.IsNumericBased(Type.NumericPersuation.Int)) {
+        // already checked that BigOrdinal or real inputs are integral
+        // FIXME - are there subset range checks to make?}
+      } else if (toType.IsNumericBased(Type.NumericPersuation.Real)) {
+        // already checked that BigOrdinal is integral
+        // FIXME - are there subset range checks to make?
```

Comment:
What is the status of these?

Reply:
Added tests for subset type. Removed comments.